File: src/main/java/org/trp/shincolle/init/ModDataComponents.java
package org.trp.shincolle.init;

import net.minecraft.core.component.DataComponentType;
import net.minecraft.core.registries.Registries;
import net.neoforged.neoforge.fluids.SimpleFluidContent;
import net.neoforged.neoforge.registries.DeferredHolder;
import net.neoforged.neoforge.registries.DeferredRegister;
import org.trp.shincolle.Shincolle;

public final class ModDataComponents {
    public static final DeferredRegister.DataComponents DATA_COMPONENTS =
            DeferredRegister.createDataComponents(Registries.DATA_COMPONENT_TYPE, Shincolle.MODID);

    public static final DeferredHolder<DataComponentType<?>, DataComponentType<SimpleFluidContent>> SHIPTANK_FLUID =
            DATA_COMPONENTS.registerComponentType("shiptank_fluid", builder -> builder
                    .persistent(SimpleFluidContent.CODEC)
                    .networkSynchronized(SimpleFluidContent.STREAM_CODEC));

    public static final DeferredHolder<DataComponentType<?>, DataComponentType<Integer>> BOOK_PAGE =
            DATA_COMPONENTS.registerComponentType("book_page", builder -> builder
                    .persistent(net.minecraft.util.ExtraCodecs.NON_NEGATIVE_INT)
                    .networkSynchronized(net.minecraft.network.codec.ByteBufCodecs.VAR_INT));

    public static final DeferredHolder<DataComponentType<?>, DataComponentType<Integer>> BOOK_CHAPTER =
            DATA_COMPONENTS.registerComponentType("book_chapter", builder -> builder
                    .persistent(net.minecraft.util.ExtraCodecs.NON_NEGATIVE_INT)
                    .networkSynchronized(net.minecraft.network.codec.ByteBufCodecs.VAR_INT));

    private ModDataComponents() {
    }
}

File: src/main/java/org/trp/shincolle/init/ModDataComponents.kt
package org.trp.shincolle.init

import net.minecraft.core.component.DataComponentType
import net.minecraft.core.registries.Registries
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.util.ExtraCodecs
import net.neoforged.neoforge.fluids.SimpleFluidContent
import net.neoforged.neoforge.registries.DeferredHolder
import net.neoforged.neoforge.registries.DeferredRegister
import org.trp.shincolle.Shincolle
import java.util.function.UnaryOperator

object ModDataComponents {
    val DATA_COMPONENTS: DeferredRegister.DataComponents =
        DeferredRegister.createDataComponents(Registries.DATA_COMPONENT_TYPE, Shincolle.MODID)

    @JvmField
    val SHIPTANK_FLUID: DeferredHolder<DataComponentType<*>?, DataComponentType<SimpleFluidContent?>?> =
        DATA_COMPONENTS.registerComponentType<SimpleFluidContent?>(
            "shiptank_fluid",
            UnaryOperator { builder: DataComponentType.Builder<SimpleFluidContent?>? ->
                builder!!
                    .persistent(SimpleFluidContent.CODEC)
                    .networkSynchronized(SimpleFluidContent.STREAM_CODEC)
            })

    val BOOK_PAGE: DeferredHolder<DataComponentType<*>?, DataComponentType<Int?>?> =
        DATA_COMPONENTS.registerComponentType<Int?>(
            "book_page",
            UnaryOperator { builder: DataComponentType.Builder<Int?>? ->
                builder!!
                    .persistent(ExtraCodecs.NON_NEGATIVE_INT)
                    .networkSynchronized(ByteBufCodecs.VAR_INT)
            })

    val BOOK_CHAPTER: DeferredHolder<DataComponentType<*>?, DataComponentType<Int?>?> =
        DATA_COMPONENTS.registerComponentType<Int?>(
            "book_chapter",
            UnaryOperator { builder: DataComponentType.Builder<Int?>? ->
                builder!!
                    .persistent(ExtraCodecs.NON_NEGATIVE_INT)
                    .networkSynchronized(ByteBufCodecs.VAR_INT)
            })
}
